R/zchunk_L222.en_transformation.R

```diff
@@ -81,248 +87,540 @@ module_energy_L222.en_transformation_DISABLED <- function(command, ...) {
+      select(-year.x) %>%
+      rename(year = year.y) %>%
+      bind_rows(L222.globaltech_retirement_base) ->
+      L222.globaltech_retirement
+
+    # Retirement may consist of any of three types of retirement function (phased, s-curve, or none)
+    # This section checks L222.globaltech_retirement for each of these functions and creates a separate level 2 file for each
+    # All of these options have different headers, and all are allowed
+    if (any(!is.na(L222.globaltech_retirement$shutdown.rate))){
+      L222.globaltech_retirement %>%
+        filter(!is.na(L222.globaltech_retirement$shutdown.rate)) %>%
+        select(sector.name, subsector.name, technology, lifetime, shutdown.rate) ->
+        L222.GlobalTechShutdown_en
+    }
+
+    if (any(!is.na(L222.globaltech_retirement$half.life))){
+      L222.globaltech_retirement %>%
+        filter(!is.na(L222.globaltech_retirement$half.life)) %>%
+        select(sector.name, subsector.name, technology, lifetime, steepness, half.life) ->
+        L222.GlobalTechSCurve_en
+    }
+
+    # L222.GlobalTechLifetime_en: Global tech lifetime
+    if (any(is.na(L222.globaltech_retirement$shutdown.rate) & is.na(L222.globaltech_retirement$half.life))){
+      L222.globaltech_retirement %>%
+        filter(is.na(L222.globaltech_retirement$shutdown.rate) & is.na(L222.globaltech_retirement$half.life)) %>%
+        select(sector.name, subsector.name, technology, lifetime) ->
+        L222.GlobalTechLifetime_en
+    }
+
+    # L222.GlobalTechProfitShutdown_en: Global tech profit shutdown decider and parameters
+    if (any(!is.na(L222.globaltech_retirement$median.shutdown.point))){
+      L222.globaltech_retirement %>%
+        filter(!is.na(L222.globaltech_retirement$median.shutdown.point)) %>%
+        select(sector.name, subsector.name, technology, median.shutdown.point, profit.shutdown.steepness) ->
+        L222.GlobalTechProfitShutdown_en
+    }
+
+    #2d. Calibration and region-specific data
+    #  generate base year calibrated outputs of gas processing by interpolating from historical values
+    L122.out_EJ_R_gasproc_F_Yh %>%
+      complete(nesting(GCAM_region_ID, sector, fuel), year = c(year, BASE_YEARS)) %>%
+      arrange(GCAM_region_ID, year) %>%
+      group_by(GCAM_region_ID, sector, fuel) %>%
+      mutate(value = approx_fun(as.numeric(year), value)) %>%
+      ungroup() %>%
+      filter(year %in% BASE_YEARS) %>%
+      # append region names
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") ->
+      L222.out_EJ_R_gasproc_F_Yh_base
+
+    # append matching calibrated technology sector/subsector/technology to calibrated outputs of gas processing
+    calibrated_techs %>%
+      semi_join(L222.out_EJ_R_gasproc_F_Yh_base, by = c("sector", "fuel")) %>%
+      select(sector, fuel, supplysector, subsector, technology) %>%
+      left_join(L222.out_EJ_R_gasproc_F_Yh_base, by = c("sector", "fuel")) %>%
```

Comment:
Can't use `left_join_error_no_match`?

Reply:
I tried it first and it failed.

Reply:
So please add a comment explaining why: usually, because there are NA values in the incoming data frame (which we expect); or, less frequently, there are multiple matches, and we expect that too.